internal/controller/scheduler/batch_buildkite_job_checker.go

```diff
@@ -0,0 +1,173 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+// BatchBuildkiteJobChecker monitors Buildkite jobs for cancellation state changes.
+// Unlike the old legacy BuildkiteJobChecker, this checker check all pending jobs together,
+// relying on the new Stack API.
+type BatchBuildkiteJobChecker struct {
+	logger      *zap.Logger
+	agentClient *api.AgentClient
+	k8s         kubernetes.Interface
+
+	// The job cancel checkers query the job state every so often.
+	jobCancelCheckerInterval time.Duration
+
+	// Store jobs that we will check against.
+	checkingJobsMu sync.Mutex
+	checkingJobs   map[uuid.UUID]metav1.ObjectMeta
+}
+
+// NewBatchBuildkiteJobChecker creates a new Batch Buildkite job checker.
+func NewBatchBuildkiteJobChecker(
+	logger *zap.Logger,
+	agentClient *api.AgentClient,
+	k8s kubernetes.Interface,
+	interval time.Duration,
+) *BatchBuildkiteJobChecker {
+	return &BatchBuildkiteJobChecker{
+		logger:                   logger,
+		agentClient:              agentClient,
+		k8s:                      k8s,
+		jobCancelCheckerInterval: interval,
+		checkingJobs:             make(map[uuid.UUID]metav1.ObjectMeta),
+	}
+}
+
+// StartChecking starts a gorouting loop to periodically check job states for all jobs that it manages.
+func (c *BatchBuildkiteJobChecker) StartChecking(ctx context.Context) {
+	go c.batchChecker(ctx)
+}
+
+func (c *BatchBuildkiteJobChecker) batchChecker(ctx context.Context) {
+	c.logger.Debug("Starting batch job state checker")
+	defer c.logger.Debug("Stopped batch job state checker")
+
+	ticker := time.NewTicker(c.jobCancelCheckerInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			c.checkJobStates(ctx)
+		}
+	}
+}
+
+func (c *BatchBuildkiteJobChecker) checkJobStates(ctx context.Context) {
+	c.checkingJobsMu.Lock()
+	if len(c.checkingJobs) == 0 {
+		c.checkingJobsMu.Unlock()
+		return
+	}
+
+	// Create slices for job UUIDs and corresponding pod metadata
+	jobUUIDs := make([]string, 0, len(c.checkingJobs))
+	jobToPodMeta := make(map[string]metav1.ObjectMeta, len(c.checkingJobs))
+
+	for jobUUID, podMeta := range c.checkingJobs {
+		jobUUIDStr := jobUUID.String()
+		jobUUIDs = append(jobUUIDs, jobUUIDStr)
+		jobToPodMeta[jobUUIDStr] = podMeta
+	}
+	c.checkingJobsMu.Unlock() // Release the lock as soon as possible.
+
+	// Split jobs into batches of 1000 and process concurrently
+	const batchSize = 1000
+	var wg sync.WaitGroup
+
+	batchCount := int(math.Ceil(float64(len(jobUUIDs)) / float64(batchSize)))
+	jobStatesCh := make(chan map[string]api.JobState, batchCount)
+
+	for batch := range slices.Chunk(jobUUIDs, batchSize) {
+		wg.Add(1)
+		go func(batchUUIDs []string) {
+			defer wg.Done()
+
+			jobStates, _, err := c.agentClient.GetJobStates(ctx, batchUUIDs)
+			if err != nil {
+				c.logger.Error("Couldn't fetch states of jobs", zap.Error(err), zap.Int("batch_size", len(batchUUIDs)))
+				return
+			}
+
+			jobStatesCh <- jobStates
+		}(batch)
```

Comment:
not a blocker:
```suggestion
		go func() {
			defer wg.Done()

			jobStates, _, err := c.agentClient.GetJobStates(ctx, batch)
			if err != nil {
				c.logger.Error("Couldn't fetch states of jobs", zap.Error(err), zap.Int("batch_size", len(batch)))
				return
			}

			jobStatesCh <- jobStates
		}()
```
i think post-Go 1.23 we don't need to make this a function arg anymore